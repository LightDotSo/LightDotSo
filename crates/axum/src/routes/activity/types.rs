// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::activity;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Activity root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Activity {
    /// The id of the activity.
    id: String,
    /// The entity id of the activity.
    entity: String,
    /// The operation type of the activity.
    operation: String,
    /// The timestamp of the activity.
    timestamp: String,
    /// The actor wallet address of the activity.
    address: Option<String>,
}

/// Implement From<activity::Data> for Activity.
impl From<activity::Data> for Activity {
    fn from(activity: activity::Data) -> Self {
        Self {
            id: activity.id,
            entity: activity.entity.to_string(),
            operation: activity.operation.to_string(),
            timestamp: activity.timestamp.to_rfc3339(),
            address: activity.wallet_address.map(|addr| addr.to_string()),
        }
    }
}
