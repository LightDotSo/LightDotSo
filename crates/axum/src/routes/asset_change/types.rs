// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::routes::{interpretation_action::types::InterpretationAction, token::types::Token};
use lightdotso_prisma::asset_change;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// AssetChange root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct AssetChange {
    /// The id of the asset change.
    id: String,
    /// The address of the asset change.
    address: String,
    /// The amount of the asset change.
    amount: i64,
    /// The before amount of the asset change.
    before_amount: Option<i64>,
    /// The after amount of the asset change.
    after_amount: Option<i64>,
    /// The action id of the asset change.
    action: Option<InterpretationAction>,
    /// The token id of the asset change.
    token: Option<Token>,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<asset_change::Data> for AssetChange.
impl From<asset_change::Data> for AssetChange {
    fn from(asset_change: asset_change::Data) -> Self {
        Self {
            id: asset_change.id,
            address: asset_change.address,
            amount: asset_change.amount.parse().unwrap_or_default(),
            before_amount: asset_change.before_amount.and_then(|s| s.parse::<i64>().ok()),
            after_amount: asset_change.after_amount.and_then(|s| s.parse::<i64>().ok()),
            action: asset_change.interpretation_action.and_then(|maybe_action| {
                maybe_action.map(|action| InterpretationAction::from(*action))
            }),
            token: asset_change
                .token
                .and_then(|maybe_token| maybe_token.map(|token| Token::from(*token))),
        }
    }
}
