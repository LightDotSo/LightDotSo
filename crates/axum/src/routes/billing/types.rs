// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::billing;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Billing root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Billing {
    /// The id of the billing.
    id: String,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing::Data> for Billing.
impl From<billing::Data> for Billing {
    fn from(billing: billing::Data) -> Self {
        Self { id: billing.id }
    }
}
