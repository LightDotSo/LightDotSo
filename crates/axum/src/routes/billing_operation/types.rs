// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lightdotso_prisma::billing_operation;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// BillingOperation root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct BillingOperation {
    /// The id of the billing operation.
    id: String,
    /// The status of the billing operation.
    status: String,
    /// The balance USD of the billing operation.
    balance_usd: f64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing_operation::Data> for BillingOperation.
impl From<billing_operation::Data> for BillingOperation {
    fn from(billing_operation: billing_operation::Data) -> Self {
        Self {
            id: billing_operation.id,
            status: billing_operation.status.to_string(),
            balance_usd: billing_operation.balance_usd,
        }
    }
}
