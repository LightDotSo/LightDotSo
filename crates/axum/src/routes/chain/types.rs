// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::chain;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Chain root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Chain {
    /// The id of the protocol group.
    id: i64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<chain::Data> for Chain.
impl From<chain::Data> for Chain {
    fn from(chain: chain::Data) -> Self {
        Self { id: chain.id }
    }
}
