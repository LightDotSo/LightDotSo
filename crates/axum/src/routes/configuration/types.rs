// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::{configuration, owner};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Configuration root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct Configuration {
    /// The id of the configuration.
    id: String,
    /// The address of the configuration.
    address: String,
    /// The image hash of the configuration.
    image_hash: String,
    /// The checkpoint of the configuration.
    checkpoint: i64,
    /// The threshold of the configuration.
    threshold: i64,
    /// The owners of the configuration.
    owners: Vec<ConfigurationOwner>,
}

/// Configuration Owner.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct ConfigurationOwner {
    /// The id of the owner.
    id: String,
    /// The index of the owner.
    index: i32,
    /// The address of the owner.
    address: String,
    /// The weight of the owner.
    weight: i64,
}

/// Implement From<configuration::Data> for Configuration.
impl From<configuration::Data> for Configuration {
    fn from(configuration: configuration::Data) -> Self {
        Self {
            id: configuration.id.to_string(),
            address: configuration.address.to_string(),
            image_hash: configuration.image_hash.to_string(),
            checkpoint: configuration.checkpoint,
            threshold: configuration.threshold,
            owners: configuration.owners.map_or(Vec::new(), |owners| {
                owners.into_iter().map(ConfigurationOwner::from).collect()
            }),
        }
    }
}

/// Implement From<owner::Data> for Owner.
impl From<owner::Data> for ConfigurationOwner {
    fn from(owner: owner::Data) -> Self {
        Self {
            id: owner.id.to_string(),
            address: owner.address.to_string(),
            index: owner.index,
            weight: owner.weight,
        }
    }
}
