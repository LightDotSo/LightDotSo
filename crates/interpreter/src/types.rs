// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::constants::InterpretationActionType;
use ethers_main::{
    abi::{Address, Uint},
    types::{Bytes, Log},
};
use foundry_evm::{trace::node::CallTraceNode, CallKind};
use revm::interpreter::InstructionResult;
use serde::{Deserialize, Serialize};

// Entire file is derived from https://github.com/EnsoFinance/transaction-simulator/blob/42bc679fb171de760838457820d5c6622e53ab15/src/simulation.rs
// License: MIT

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InterpretationRequest {
    /// Block number of the request
    pub block_number: Option<u64>,
    /// Gas limit of the transaction
    pub gas_limit: u64,
    /// From address of the transaction
    pub from: Address,
    /// To address of the transaction
    pub to: Address,
    /// Chain ID of the simulation
    pub chain_id: u64,
    /// Call data of the transaction
    pub call_data: Option<Bytes>,
    /// Value to send
    pub value: Option<u64>,
    /// Trace of the transaction
    pub traces: Vec<CallTrace>,
    /// Logs of the transaction
    pub logs: Vec<Log>,
}

impl Default for InterpretationRequest {
    fn default() -> Self {
        Self {
            block_number: None,
            gas_limit: u64::MAX,
            from: Address::default(),
            to: Address::default(),
            chain_id: 0,
            call_data: Some(Bytes::new()),
            value: None,
            traces: Vec::new(),
            logs: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InterpretationResponse {
    /// Gas used by the transaction
    pub gas_used: u64,
    /// Block number of the simulation
    pub block_number: u64,
    /// Whether the transaction was successful
    pub success: bool,
    /// Trace of the transaction
    pub traces: Vec<CallTrace>,
    /// Logs of the transaction
    pub logs: Vec<Log>,
    /// Exit reason of the transaction
    pub exit_reason: InstructionResult,
    /// Actions that were interpreted
    pub actions: Vec<InterpretationAction>,
    /// Changes in the assets of the transaction
    pub asset_changes: Vec<AssetChange>,
}

impl Default for InterpretationResponse {
    fn default() -> Self {
        Self {
            gas_used: 0,
            block_number: 0,
            success: false,
            traces: Vec::new(),
            logs: Vec::new(),
            exit_reason: InstructionResult::Stop,
            actions: Vec::new(),
            asset_changes: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InterpretationAction {
    pub action_type: InterpretationActionType,
    pub address: Option<Address>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AdapterResponse {
    /// Actions that were interpreted
    pub actions: Vec<InterpretationAction>,
    /// Changes in the assets of the transaction
    pub asset_changes: Vec<AssetChange>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AssetChange {
    pub address: Address,
    pub before_amount: Uint,
    pub after_amount: Uint,
    pub amount: Uint,

    pub action: InterpretationAction,
    pub token: AssetToken,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AssetToken {
    pub address: Address,
    pub token_id: Option<Uint>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CallTrace {
    pub call_type: CallKind,
    pub from: Address,
    pub to: Address,
    pub value: Option<Uint>,
}

impl From<CallTraceNode> for CallTrace {
    fn from(node: CallTraceNode) -> Self {
        CallTrace {
            call_type: node.kind(),
            from: node.trace.caller,
            to: node.trace.address,
            value: Some(node.trace.value),
        }
    }
}
