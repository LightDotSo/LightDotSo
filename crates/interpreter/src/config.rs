// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::{
    interpreter::Interpreter,
    types::{InterpretationRequest, InterpretationResponse},
};
use clap::Parser;
use eyre::Result;
use lightdotso_simulator::types::SimulationRequest;
use lightdotso_tracing::tracing::info;

#[derive(Debug, Clone, Parser, Default)]
pub struct InterpreterArgs {}

impl InterpreterArgs {
    pub async fn run(self, requests: Vec<SimulationRequest>) -> Result<InterpretationResponse> {
        // Add info
        info!("InterpreterArgs run, starting...");

        // Print the config
        info!("Config: {:?}", self);

        // Construct the interpreter
        let mut interpreter = Interpreter::new().await;

        info!("InterpreterArgs run, starting simulate...");

        // Simulate the user operation
        let res = interpreter.run_with_simulate_bundle(requests).await?;

        info!("res: {:?}", res);

        info!("InterpreterArgs run, finished");

        Ok(res)
    }

    pub async fn run_interpretation(
        self,
        request: InterpretationRequest,
    ) -> Result<InterpretationResponse> {
        // Add info
        info!("InterpreterArgs run, starting...");

        // Print the config
        info!("Config: {:?}", self);

        // Construct the interpreter
        let mut interpreter = Interpreter::new().await;

        info!("InterpreterArgs run, starting interpret...");

        // Interpret the user operation
        let res = interpreter.run_with_interpret(request).await?;

        info!("res: {:?}", res);

        info!("InterpreterArgs run, finished");

        Ok(res)
    }
}
