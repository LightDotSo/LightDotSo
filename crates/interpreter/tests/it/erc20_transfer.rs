// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use clap::Parser;
use eyre::Result;
use lightdotso_interpreter::config::InterpreterArgs;
use lightdotso_simulator::types::SimulationRequest;

// https://etherscan.io/tx/0xee623726751e879ca379d3680a7658e307a6cbc3aa99be7f2706470eebdd969d

#[tokio::test(flavor = "multi_thread")]
async fn test_integration_erc20_transfer() -> Result<()> {
    let request = SimulationRequest {
        chain_id: 1,
        // kaki.eth
        from: "0x4fd9D0eE6D6564E80A9Ee00c0163fC952d0A45Ed".parse()?,
        // ENS token address
        to: "0xC18360217D8F7Ab5e7c516566761Ea12Ce7F9D72".parse()?,
        data: Some("0xa9059cbb0000000000000000000000006f8a90995fdce00da1f7dd731d812f6a6d18d1ff000000000000000000000000000000000000000000000001a055690d9db80000".parse()?),
        value: None,
        gas_limit: u64::MAX,
        // Tx was on 13704035
        block_number: Some(13704034),
    };

    // Parse the command line arguments
    let args = InterpreterArgs::parse_from([""]);

    // Run the interpreter
    let res = args.run(request).await?;

    assert!(res.success);

    println!("Result: {:?}", res);

    Ok(())
}
