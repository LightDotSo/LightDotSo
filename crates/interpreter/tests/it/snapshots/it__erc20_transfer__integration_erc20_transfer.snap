---
source: crates/interpreter/tests/it/erc20_transfer.rs
expression: res
---
InterpretationResponse {
    gas_used: 85889,
    block_number: 13704034,
    success: true,
    traces: [
        CallTrace {
            call_type: Call,
            from: 0x4fd9d0ee6d6564e80a9ee00c0163fc952d0a45ed,
            to: 0xc18360217d8f7ab5e7c516566761ea12ce7f9d72,
            value: Some(
                0,
            ),
        },
    ],
    logs: [
        Log {
            address: 0xc18360217d8f7ab5e7c516566761ea12ce7f9d72,
            topics: [
                0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef,
                0x0000000000000000000000004fd9d0ee6d6564e80a9ee00c0163fc952d0a45ed,
                0x0000000000000000000000006f8a90995fdce00da1f7dd731d812f6a6d18d1ff,
            ],
            data: Bytes(0x000000000000000000000000000000000000000000000001a055690d9db80000),
            block_hash: None,
            block_number: None,
            transaction_hash: None,
            transaction_index: None,
            log_index: None,
            transaction_log_index: None,
            log_type: None,
            removed: None,
        },
        Log {
            address: 0xc18360217d8f7ab5e7c516566761ea12ce7f9d72,
            topics: [
                0xdec2bacdd2f05b59de34da9b523dff8be42e5e38e818c82fdb0bae774387a724,
                0x0000000000000000000000005bfcb4be4d7b43437d5a0c57e908c048a4418390,
            ],
            data: Bytes(0x000000000000000000000000000000000000000000002c0c5f545fac2d32d119000000000000000000000000000000000000000000002c0abefef69e8f7ad119),
            block_hash: None,
            block_number: None,
            transaction_hash: None,
            transaction_index: None,
            log_index: None,
            transaction_log_index: None,
            log_type: None,
            removed: None,
        },
    ],
    exit_reason: Return,
    formatted_trace: "  [64245] \u{1b}[32m0xC18360217D8F7Ab5e7c516566761Ea12Ce7F9D72\u{1b}[0m::\u{1b}[32mtransfer\u{1b}[0m(0x6f8A90995FdcE00da1F7dd731d812F6a6D18d1ff, 30000000000000000000 \u{1b}[2;49;39m[3e19]\u{1b}[0m) \u{1b}[33m\u{1b}[0m\n    ├─ emit \u{1b}[36mTransfer\u{1b}[0m(param0: 0x4fd9D0eE6D6564E80A9Ee00c0163fC952d0A45Ed, param1: 0x6f8A90995FdcE00da1F7dd731d812F6a6D18d1ff, param2: 30000000000000000000 \u{1b}[2;49;39m[3e19]\u{1b}[0m)\n    ├─ emit \u{1b}[36mDelegateVotesChanged\u{1b}[0m(param0: 0x5BFCB4BE4d7B43437d5A0c57E908c048a4418390, param1: 208012355395673648189721 \u{1b}[2;49;39m[2.08e23]\u{1b}[0m, param2: 207982355395673648189721 \u{1b}[2;49;39m[2.079e23]\u{1b}[0m)\n    └─ \u{1b}[32m← \u{1b}[0m0x0000000000000000000000000000000000000000000000000000000000000001\n",
    actions: [
        InterpretationAction {
            action_type: ERC20Send,
            address: Some(
                0x4fd9d0ee6d6564e80a9ee00c0163fc952d0a45ed,
            ),
        },
        InterpretationAction {
            action_type: ERC20Receive,
            address: Some(
                0x6f8a90995fdce00da1f7dd731d812f6a6d18d1ff,
            ),
        },
    ],
    asset_changes: [
        AssetChange {
            address: 0x4fd9d0ee6d6564e80a9ee00c0163fc952d0a45ed,
            before_amount: 30000000000000000000,
            after_amount: 0,
            amount: 30000000000000000000,
            action: InterpretationAction {
                action_type: ERC20Send,
                address: Some(
                    0x4fd9d0ee6d6564e80a9ee00c0163fc952d0a45ed,
                ),
            },
            token: AssetToken {
                address: 0xc18360217d8f7ab5e7c516566761ea12ce7f9d72,
                token_id: None,
            },
        },
        AssetChange {
            address: 0x6f8a90995fdce00da1f7dd731d812f6a6d18d1ff,
            before_amount: 0,
            after_amount: 30000000000000000000,
            amount: 30000000000000000000,
            action: InterpretationAction {
                action_type: ERC20Receive,
                address: Some(
                    0x6f8a90995fdce00da1f7dd731d812f6a6d18d1ff,
                ),
            },
            token: AssetToken {
                address: 0xc18360217d8f7ab5e7c516566761ea12ce7f9d72,
                token_id: None,
            },
        },
    ],
}
